onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -4,213 +4,577 @@
+common::Status GetSizeInBytesFromTensorProto(const ONNX_NAMESPACE::TensorProto& tensor_proto, size_t* out) {
+  const auto& dims = tensor_proto.dims();
+  size_t size = 1;
+  for (int i = 0; i < dims.size(); ++i) {
+    if (dims[i] < 0) {
+      return common::Status(common::ONNXRUNTIME, common::FAIL, "Invalid TensorProto");
+    }
+    if (!IAllocator::CalcMemSizeForArray(size, static_cast<size_t>(dims[i]), &size)) {
+      return common::Status(common::ONNXRUNTIME, common::FAIL, "Invalid TensorProto");
+    }
+  }
+  switch (tensor_proto.data_type()) {
+    CASE_PROTO_TRACE(FLOAT, float);
+    CASE_PROTO_TRACE(DOUBLE, double);
+    CASE_PROTO_TRACE(BOOL, bool);
+    CASE_PROTO_TRACE(INT8, int8_t);
+    CASE_PROTO_TRACE(INT16, int16_t);
+    CASE_PROTO_TRACE(INT32, int32_t);
+    CASE_PROTO_TRACE(INT64, int64_t);
+    CASE_PROTO_TRACE(UINT8, uint8_t);
+    CASE_PROTO_TRACE(UINT16, uint16_t);
+    CASE_PROTO_TRACE(UINT32, uint32_t);
+    CASE_PROTO_TRACE(UINT64, uint64_t);
+    CASE_PROTO_TRACE(FLOAT16, MLFloat16);
+    CASE_PROTO_TRACE(BFLOAT16, BFloat16);
+    CASE_PROTO_TRACE(STRING, std::string);
+    default:
+      return common::Status(common::ONNXRUNTIME, common::NOT_IMPLEMENTED);
+  }
+  return Status::OK();
+}
+
+std::vector<int64_t> GetTensorShapeFromTensorShapeProto(const ONNX_NAMESPACE::TensorShapeProto& tensor_shape_proto) {
+  const auto& dims = tensor_shape_proto.dim();
+  std::vector<int64_t> tensor_shape_vec(static_cast<size_t>(dims.size()));
+  for (int i = 0; i < dims.size(); ++i) {
+    tensor_shape_vec[i] = dims[i].has_dim_param() ? -1 /* symbolic dimensions are represented as -1 in onnxruntime*/
+                                                  : dims[i].dim_value();
+  }
+  return tensor_shape_vec;
+}
+
+template <typename T>
+common::Status GetTensorByTypeFromTensorProto(const TensorProto& tensor_proto, const TensorShape& tensor_shape,
+                                              std::unique_ptr<Tensor>* p_tensor, const OrtAllocatorInfo& alloc,
+                                              const void* raw_data, size_t raw_data_len, void* preallocated,
+                                              size_t preallocated_size) {
   int64_t tensor_size = tensor_shape.Size();
-  if (tensor_size < 0) {
-    return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT, "Tensor shape cannot contain any negative value");
+  // tensor_size could be zero. see test_slice_start_out_of_bounds\test_data_set_0\output_0.pb
+  if (tensor_size < 0 || static_cast<uint64_t>(tensor_size) > SIZE_MAX) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Invalid shape ", tensor_shape);
   }
-  static_assert(sizeof(BFloat16) == sizeof(uint16_t), "BFloat16 must has 16 bit size");
   size_t size_to_allocate;
-  if (!IAllocator::CalcMemSizeForArrayWithAlignment<256>(static_cast<size_t>(tensor_size), sizeof(BFloat16), &size_to_allocate)) {
+  if (!IAllocator::CalcMemSizeForArrayWithAlignment<0>(static_cast<size_t>(tensor_size), sizeof(T),
```

Comment:
>CalcMemSizeForArrayWithAlignment<0>( [](start = 19, length = 36)

Should this use alignment of 256 if preallocated is nullptr?

Reply:
No. Here it won't do the allocation.  It just check if the buffer has enough size

---
In reply to: [261419514](https://github.com/Microsoft/onnxruntime/pull/520#discussion_r261419514) [](ancestors = 261419514)